### PR TITLE
Payment channel actor

### PR DESCRIPTION
**Pull request by wer1st**
### Description of the Change
Implementation of payment channel actor v0, v2, v3.
Implementation of system actor v2, v3.
Implementation of account actor v3.

### CODE REVIEW

File: test/core/test_vectors/test.cpp [hunk 1]
```diff
@@ -256,11 +256,23 @@ auto search() {
           kCorpusRoot + "/extracted/0004-coverage-boost/fil_1_storageminer/SubmitWindowedPoSt/Ok/ext-0004-fil_1_storageminer-SubmitWindowedPoSt-Ok-9.json"
       };
 
+      // Skip tests that fail in Fuhon
+      // TODO (@wer1st) these tests should be enabled after FIL-260
```
Comment (Alexey-N-Chernyshov) [inline test/core/test_vectors/test.cpp:260]: Please, add brief description of the problem, so developers haven't search for FIL-260 every time they see the TODO.

**Review by Alexey-N-Chernyshov — APPROVED** [review]
Nice!

File: test/core/test_vectors/test.cpp (hunk 1 above)
Comment (wer1st) [reply test/core/test_vectors/test.cpp:260]: done

File: core/vm/runtime/env.hpp [hunk 2]
```diff
@@ -21,9 +21,19 @@ namespace fc::vm::runtime {
   using state::StateTree;
   using state::StateTreeImpl;
 
+  /**
+   * Returns the public key type of address (`BLS`/`SECP256K1`) of an account
+   * actor identified by `address`.
+   * @param state_tree - state tree
+   * @param ipld - regular or charging ipld
+   * @param address - account actor address
+   * @param is_actor - true if address is an actor
```
Comment (turuslan) [inline core/vm/runtime/env.hpp:30]:
```suggestion
   * @param allow_actor - is actor type address allowed
```

File: core/vm/actor/builtin/v0/payment_channel/payment_channel_actor.hpp [hunk 3]
```diff
@@ -3,18 +3,21 @@
...
 #include "primitives/address/address_codec.hpp"
 #include "vm/actor/actor_method.hpp"
 #include "vm/actor/builtin/v0/payment_channel/payment_channel_actor_state.hpp"
 
 namespace fc::vm::actor::builtin::v0::payment_channel {
   using primitives::EpochDuration;
 
-  constexpr size_t kLaneLimit{256};
-  constexpr EpochDuration kSettleDelay{1};
+  constexpr size_t kLaneLimit{LLONG_MAX};
```
Comment (turuslan) [inline core/vm/actor/builtin/v0/payment_channel/payment_channel_actor.hpp:18]:
```suggestion
  constexpr size_t kLaneLimit{INT64_MAX};
```

File: core/vm/actor/builtin/v0/payment_channel/payment_channel_actor_state.hpp [hunk 4]
```diff
@@ -112,7 +105,18 @@ namespace fc::vm::actor::builtin::v0::payment_channel {
     Buffer extra;
     Buffer proof;
   };
-  CBOR_TUPLE(PaymentVerifyParams, extra, proof);
+  CBOR_TUPLE(PaymentVerifyParams, extra);
```
Comment (turuslan) [inline core/vm/actor/builtin/v0/payment_channel/payment_channel_actor_state.hpp:108]: [0.9.13 has this field](https://github.com/filecoin-project/specs-actors/blob/v0.9.13/actors/builtin/paych/paych_actor.go#L141)

File: core/vm/actor/builtin/v2/payment_channel/payment_channel_actor.cpp [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    }
+    if (!voucher.secret_preimage.empty()
+        && gsl::make_span(runtime.hashBlake2b(params.secret))
+               != gsl::make_span(voucher.secret_preimage)) {
+      return VMExitCode::kErrIllegalArgument;
+    }
+    if (voucher.extra) {
+      OUTCOME_TRY(params_extra, encodeActorParams(voucher.extra->params));
+      OUTCOME_TRY(runtime.send(
+          voucher.extra->actor, voucher.extra->method, params_extra, 0));
```
Comment (turuslan) [inline core/vm/actor/builtin/v2/payment_channel/payment_channel_actor.cpp:96]:
```suggestion
      OUTCOME_TRY(runtime.send(
          voucher.extra->actor, voucher.extra->method, voucher.extra->params, 0));
```
https://github.com/filecoin-project/specs-actors/blob/master/actors/builtin/paych/paych_actor.go#L216

File: core/vm/runtime/impl/runtime_impl.cpp [hunk 6]
```diff
@@ -110,11 +112,50 @@ namespace fc::vm::runtime {
...
-  fc::outcome::result<void> RuntimeImpl::deleteActor() {
-    // TODO: transfer to kBurntFundsActorAddress
-    // TODO(a.chernyshov) FIL-137 implement state_tree remove if needed
-    // return state_tree_->remove(address);
-    return fc::outcome::failure(RuntimeError::kUnknown);
+  fc::outcome::result<void> RuntimeImpl::deleteActor(const Address &address) {
+    OUTCOME_TRY(chargeGas(execution_->env->pricelist.onDeleteActor()));
+    auto &state{*execution()->state_tree};
+    if (auto _actor{state.get(getCurrentReceiver())}) {
+      if (_actor.has_error() && _actor.error() == HamtError::kNotFound) {
```
Comment (turuslan) [inline core/vm/runtime/impl/runtime_impl.cpp:119]: `if (auto _actor{error})` will go to `else`, so `if (_actor.has_error() && ...)` is always false. Move to `else` or remove

File: core/vm/runtime/runtime.hpp [hunk 7]
```diff
@@ -137,9 +137,22 @@ namespace fc::vm::runtime {
...
+    virtual outcome::result<void> deleteActor(const Address &address) = 0;
+
+    /**
+     * @brief Transfer debits money from one account and credits it to another
+     *
+     * @param debitFrom - money sender
+     * @param creditTo - money receiver
+     * @param amount - amount of money that is transfered
+     */
+    virtual outcome::result<void> transfer(const Address &debitFrom,
```
Comment (turuslan) [inline core/vm/runtime/runtime.hpp:153]: Nobody uses `transfer` (except `RuntimeImpl`). Should it be virtual?

File: core/vm/runtime/runtime.hpp [hunk 8]
```diff
@@ -137,9 +137,22 @@ namespace fc::vm::runtime {
     /**
      * @brief Deletes an actor in the state tree
      *
+     * @param address - Address of actor that should be deleted
```
Comment (turuslan) [inline core/vm/runtime/runtime.hpp:140]:
```suggestion
     * @param address - Address of actor that receives remaining balance
```

File: test/core/vm/actor/builtin/v0/payment_channel/payment_channel_actor_test.cpp [hunk 9]
```diff
@@ -155,7 +157,9 @@ TEST_F(PaymentChannelActorTest, ConstructSuccess) {
   EXPECT_EQ(state.to_send, 0);
   EXPECT_EQ(state.settling_at, 0);
   EXPECT_EQ(state.min_settling_height, 0);
-  EXPECT_EQ(state.lanes.size(), 0);
+
+  EXPECT_OUTCOME_TRUE(lanes_size, state.lanes.size());
+  EXPECT_EQ(lanes_size, 0);
```
Comment (turuslan) [inline test/core/vm/actor/builtin/v0/payment_channel/payment_channel_actor_test.cpp:162]:
```suggestion
  EXPECT_OUTCOME_EQ(state.lanes.size(), 0);
```

**Review by turuslan — APPROVED** [review]

File: core/vm/actor/builtin/v0/payment_channel/payment_channel_actor_state.hpp (hunk 4 above)
Comment (wer1st) [reply core/vm/actor/builtin/v0/payment_channel/payment_channel_actor_state.hpp:108]: Good catch, I missed it when split the actor into versions.

File: core/vm/runtime/runtime.hpp (hunk 7 above)
Comment (wer1st) [reply core/vm/runtime/runtime.hpp:153]: Is there a possibility that some actor will use it in next versions?

File: core/vm/runtime/env.hpp (hunk 2 above)
Comment (wer1st) [reply core/vm/runtime/env.hpp]: changed

File: core/vm/actor/builtin/v0/payment_channel/payment_channel_actor.hpp (hunk 3 above)
Comment (wer1st) [reply core/vm/actor/builtin/v0/payment_channel/payment_channel_actor.hpp]: changed

File: core/vm/actor/builtin/v0/payment_channel/payment_channel_actor_state.hpp (hunk 4 above)
Comment (wer1st) [reply core/vm/actor/builtin/v0/payment_channel/payment_channel_actor_state.hpp]: fixed

File: core/vm/actor/builtin/v2/payment_channel/payment_channel_actor.cpp (hunk 5 above)
Comment (wer1st) [reply core/vm/actor/builtin/v2/payment_channel/payment_channel_actor.cpp]: fixed

File: core/vm/runtime/impl/runtime_impl.cpp (hunk 6 above)
Comment (wer1st) [reply core/vm/runtime/impl/runtime_impl.cpp]: fixed

File: core/vm/runtime/runtime.hpp (hunk 8 above)
Comment (wer1st) [reply core/vm/runtime/runtime.hpp]: changed

File: test/core/vm/actor/builtin/v0/payment_channel/payment_channel_actor_test.cpp (hunk 9 above)
Comment (wer1st) [reply test/core/vm/actor/builtin/v0/payment_channel/payment_channel_actor_test.cpp:162]: I couldn't. size() returns outcome::result<uint64_t> not just uint64_t. And test comparator can't compare these values.

**Review by turuslan — APPROVED** [review]

**Pull request merged**